driver-core/src/main/com/mongodb/internal/connection/ByteBufferBsonOutput.java

```diff
@@ -273,6 +276,161 @@ public void close() {
         }
     }
 
+    @Override
+    protected int writeCharacters(final String str, final boolean checkForNullCharacters) {
+        ensureOpen();
+        ByteBuf buf = getCurrentByteBuffer();
+        if ((buf.remaining() >= str.length() + 1)) {
+            if (buf.hasArray()) {
+                return writeCharactersOnArray(str, checkForNullCharacters, buf);
+            } else if (buf instanceof NettyByteBuf) {
+                return writeCharactersOnNettyByteBuf(str, checkForNullCharacters, buf);
+            }
+        }
+        return super.writeCharacters(str, 0, checkForNullCharacters);
+    }
+
+    private static void validateNoNullSingleByteChars(String str, long chars, int i) {
+        long tmp = (chars & 0x7F7F7F7F7F7F7F7FL) + 0x7F7F7F7F7F7F7F7FL;
+        tmp = ~(tmp | chars | 0x7F7F7F7F7F7F7F7FL);
+        if (tmp != 0) {
+            int firstZero = Long.numberOfTrailingZeros(tmp) >>> 3;
+            throw new BsonSerializationException(format("BSON cstring '%s' is not valid because it contains a null character "
+                    + "at index %d", str, i + firstZero));
+        }
+    }
+
+    private static void validateNoNullAsciiCharacters(String str, long asciiChars, int i) {
+        // simplified Hacker's delight search for zero with ASCII chars i.e. which doesn't use the MSB
+        long tmp = asciiChars + 0x7F7F7F7F7F7F7F7FL;
+        // MSB is 0 iff the byte is 0x00, 1 otherwise
+        tmp = ~tmp & 0x8080808080808080L;
+        // MSB is 1 iff the byte is 0x00, 0 otherwise
+        if (tmp != 0) {
+            // there's some 0x00 in the word
+            int firstZero = Long.numberOfTrailingZeros(tmp) >> 3;
+            throw new BsonSerializationException(format("BSON cstring '%s' is not valid because it contains a null character "
+                    + "at index %d", str, i + firstZero));
+        }
+    }
+
+    private int writeCharactersOnNettyByteBuf(String str, boolean checkForNullCharacters, ByteBuf buf) {
```

Comment:
I have tried hard to reduce the burden of Netty accessibility checks and the CPU cost of writing byte per byte with this, to not mention amortize the cost of branch mispredict for less common cases (non ASCII String).